sql/src/main/scala/org/apache/spark/sql/sedona_sql/expressions/raster.scala

```diff
@@ -0,0 +1,245 @@
+  }
+
+  private def readGeometry(url: String): Geometry = {
+
+    val format = GridFormatFinder.findFormat(url)
+    val hints = new Hints(Hints.FORCE_LONGITUDE_FIRST_AXIS_ORDER, true)
+    val reader = format.getReader(url, hints)
+    var coverage:GridCoverage2D = null
+
+    try coverage = reader.read(null)
+    catch {
+      case giveUp: IOException =>
+        throw new RuntimeException(giveUp)
+    }
+    reader.dispose()
+    val source = coverage.getCoordinateReferenceSystem
+    val target = CRS.decode("EPSG:4326", true)
+    val targetCRS = CRS.findMathTransform(source, target)
+    val gridRange2D = coverage.getGridGeometry.getGridRange
+    val cords = Array(Array(gridRange2D.getLow(0), gridRange2D.getLow(1)), Array(gridRange2D.getLow(0), gridRange2D.getHigh(1)), Array(gridRange2D.getHigh(0), gridRange2D.getHigh(1)), Array(gridRange2D.getHigh(0), gridRange2D.getLow(1)))
+    val polyCoordinates = new Array[Coordinate](5)
+    var index = 0
+
+    for (point <- cords) {
+      val coordinate2D = new GridCoordinates2D(point(0), point(1))
+      val result = coverage.getGridGeometry.gridToWorld(coordinate2D)
+      polyCoordinates({
+        index += 1; index - 1
+      }) = new Coordinate(result.getOrdinate(0), result.getOrdinate(1))
+    }
+
+    polyCoordinates(index) = polyCoordinates(0)
+    val factory = new GeometryFactory
+    val polygon = JTS.transform(factory.createPolygon(polyCoordinates), targetCRS)
+
+    polygon
+
+  }
+  override def dataType: DataType = GeometryUDT
+
+  override def children: Seq[Expression] = inputExpressions
+}
+
+
+// Constructs a raster dataframe from a raster image which contains multiple columns such as Geometry, Band values etc
+case class ST_DataframeFromRaster(inputExpressions: Seq[Expression])
+  extends Expression with CodegenFallback with UserDataGeneratator {
+  override def nullable: Boolean = false
+
+  override def eval(inputRow: InternalRow): Any = {
+    // This is an expression which takes one input expressions
+    assert(inputExpressions.length == 2)
+    val geomString = inputExpressions(0).eval(inputRow).asInstanceOf[UTF8String].toString
+    val totalBands = inputExpressions(1).eval(inputRow).asInstanceOf[Int]
+    val geometry = readGeometry(geomString)
+    val bandvalues = getBands(geomString, totalBands).toArray
+    returnValue(geometry.toGenericArrayData,bandvalues, 2)
+  }
+
+  private def readGeometry(url: String): Geometry = {
+
+    val format = GridFormatFinder.findFormat(url)
+    val hints = new Hints(Hints.FORCE_LONGITUDE_FIRST_AXIS_ORDER, true)
+    val reader = format.getReader(url, hints)
+    var coverage: GridCoverage2D = null
+
+    try coverage = reader.read(null)
+    catch {
+      case giveUp: IOException =>
+        throw new RuntimeException(giveUp)
+    }
+    reader.dispose()
+    val source = coverage.getCoordinateReferenceSystem
+    val target = CRS.decode("EPSG:4326", true)
+    val targetCRS = CRS.findMathTransform(source, target)
```

Comment:
If CRS transformation is not necessary here,  don't do it then.

Reply:
It looks like you are reading geom and bands together in ST_DataframeFromRaster. A better name could be ST_GeomWithBandsFromGeoTiff